File: mobile_app/services/extraordinarios.ts
// mobile_app/services/extraordinariosApi.ts
import axios from 'axios';
import { api } from './api';

const ENDPOINT_EXTRAORDINARIOS = '/api/v1/extraordinarios';

export interface ExtraordinarioItemDto {
  id: string;
  nombre: string;
  categoria_nombre: string | null;
  tipo: 'GASTO' | 'INGRESO';
  importe: number;

  pagado?: boolean | null;
  cobrado?: boolean | null;
  kpi: boolean;
  activo: boolean;

  fecha_referencia: string; // ISO datetime

  // ---- extras para navegación a forms (opcionales) ----
  periodicidad?: string | null;

  // gasto
  tipo_id?: string | null;
  segmento_id?: string | null;
  proveedor_id?: string | null;
  cuenta_id?: string | null;
  referencia_vivienda_id?: string | null;
  fecha?: string | null; // ISO date
  rango_pago?: string | null;
  comentarios?: string | null;
  importe_cuota?: number | null;
  cuotas?: number | null;

  // ingreso
  rango_cobro?: string | null;
  fecha_inicio?: string | null; // ISO date
}

export interface ExtraordinariosResponseDto {
  year: number;
  month: number; // 1-12

  total_gastos: number; // gastos extraordinarios
  total_ingresos: number; // ingresos extraordinarios

  // NUEVO
  total_gastos_omitidos: number;
  gastos_omitidos: ExtraordinarioItemDto[];

  // Balance solicitado:
  // ingresos + gastos_omitidos - gastos_extraordinarios
  balance: number;

  gastos: ExtraordinarioItemDto[];
  ingresos: ExtraordinarioItemDto[];
}

/**
 * year: año completo (ej. 2025)
 * month: 1-12
 */
export async function fetchExtraordinarios(
  year: number,
  month: number
): Promise<ExtraordinariosResponseDto> {
  const params: Record<string, any> = { year, month };

  try {
    console.log(
      '[extraordinariosApi] GET extraordinarios ->',
      ENDPOINT_EXTRAORDINARIOS,
      'params:',
      params
    );

    const res = await api.get<ExtraordinariosResponseDto>(
      ENDPOINT_EXTRAORDINARIOS,
      { params }
    );

    return res.data;
  } catch (err) {
    console.error(
      '[extraordinariosApi] Error cargando extraordinarios',
      axios.isAxiosError(err) ? err.response?.data : err
    );
    throw err;
  }
}

File: backend/app/schemas/extraordinarios.py
# backend/app/schemas/extraordinarios.py
from datetime import datetime, date
from typing import List, Optional, Literal

from pydantic import BaseModel


class ExtraordinarioItem(BaseModel):
    id: str
    nombre: str
    categoria_nombre: Optional[str] = None  # p.ej. tipo_rel.nombre
    tipo: Literal["GASTO", "INGRESO"]
    importe: float

    # Estado
    pagado: Optional[bool] = None
    cobrado: Optional[bool] = None
    kpi: bool
    activo: bool

    # Fecha de referencia para el filtro/listado:
    # - gastos extraordinarios: ultimo_pago_on
    # - ingresos extraordinarios: ultimo_ingreso_on
    # - omitidos: ultimo_omitido_on
    fecha_referencia: datetime

    # ==========================
    # Campos extra para navegar a forms
    # (opcionales para no romper nada)
    # ==========================
    periodicidad: Optional[str] = None

    # Gasto (gestionable)
    tipo_id: Optional[str] = None
    segmento_id: Optional[str] = None
    proveedor_id: Optional[str] = None
    cuenta_id: Optional[str] = None
    referencia_vivienda_id: Optional[str] = None
    fecha: Optional[date] = None
    rango_pago: Optional[str] = None
    comentarios: Optional[str] = None
    importe_cuota: Optional[float] = None
    cuotas: Optional[int] = None

    # Ingreso
    rango_cobro: Optional[str] = None
    fecha_inicio: Optional[date] = None

    class Config:
        orm_mode = True


class ExtraordinariosResponse(BaseModel):
    year: int
    month: int  # 1-12

    # Extraordinarios “clásicos” (PAGO UNICO + inactivos + pagado/cobrado)
    total_gastos: float
    total_ingresos: float

    # NUEVO: omitidos del mes (gastos)
    total_gastos_omitidos: float

    # Balance solicitado:
    # Ingresos + gastos omitidos - gastos extraordinarios
    balance: float

    gastos: List[ExtraordinarioItem]
    ingresos: List[ExtraordinarioItem]

    # NUEVO: listado de omitidos
    gastos_omitidos: List[ExtraordinarioItem]

    class Config:
        orm_mode = True